Python/bytecodes.c

```diff
@@ -1277,13 +1276,13 @@ dummy_func(
             }
         }
 
-        tier1 inst(CLEANUP_THROW, (sub_iter_st, last_sent_val_st, exc_value_st -- none, value)) {
+        tier1 inst(CLEANUP_THROW, (sub_iter_st, last_sent_val_st, exc_value_st -- none, value: _PyStackRef *)) {
```

Comment:
This doesn't seem right. `value` is a reference to an object, like any other item on the stack.

I don't think the concept of a pointer to a stack ref makes a lot of sense. Arrays, yes, pointers, no.

I'm a bit worried that the code generator does what you want here, as it is isn't what you say, or what it should do.

Reply:
Yeah I'm trying to say "give me a pointer instead of the actual thing". And it does do what I want, is that not intended?

Reply:
Although that is undesirable as it isn't clear *why* you want a single item array.
Ideally, we should mark any escaping code and let the code generator spill the stack pointer.
We will need this information anyway, if we are to implement top-of-stack caching, as that will need to spill the cached values to memory across escaping calls.

Reply:
We want a pointer to a specific stack location here. It's important that the result is assigned to the stack (and therefore visible to the GC) before anything that might trigger a GC happens.

This is not related to spilling the stack pointer or making the *stack pointer* visible to the GC>

Reply:
To expand on my above comment: A `_PyStackRef` only protects an object from destruction if it's part of the interpreter's evaluation stack or frame's local variables. A `_PyStackRef` that exists only as a local variables in C doesn't prevent object destruction. For example:

**BAD**:
```c
_PyStackRef tmp = PyStackRef_FromPyObjectNew(...);
Py_DECREF(...); // e.g., potentially trigger a GC
stack_pointer[-1] = tmp;
```

The problem is that if `tmp` is the only reference to some object, then the GC will free it because `tmp` is not visible to the GC.

**GOOD**:
```c
stack_pointer[-1] = PyStackRef_FromPyObjectNew(...);
Py_DECREF(...); // e.g., potentially trigger a GC
```

Ideally, `_PyStackRef` would only exist as part of `_PyInterpreterFrame.localsplus` and never as local variables in C, but that's awkward and hard to enforce in C.

Reply:
I am aware of all of that.

We don't want a pointer to a specific stack location.
We want the GC to be able to find all the references on the stack.

If the code generator lacks the functionality you need, then the solution is to add that functionality, not to add hacky workarounds

If the current implementation hides values on the stack from the GC, then that's a problem, but unless we explicitly mark where the GC might run (i.e. escaping calls) then there is no way for the code generator to make sure that the GC can see whole stack without keeping it in memory all the time, and that's bad for performance.

Reply:
The current implementation scans the whole stack instead of just to stacktop. There is no hiding of values.

Reply:
I addressed the review by using the correct notation in the DSL.

Reply:
@markshannon, how is `value[1]` appropriate here? There's no 1-element array here.

> We don't want a pointer to a specific stack location.

Why do you say that? We want a pointer to the location of where `value` (the result) is stored on the stack.

> If the current implementation hides values on the stack from the GC...

It's not hiding values *on the interpreter evaluation stack* from the GC. Values in C local variables are not visible to the GC.

> then there is no way for the code generator to make sure that the GC can see whole stack without keeping it in memory all the time...

What are you referring to here? The stack is currently always in memory.

Reply:
Never mind about the `value[1]` notation -- I see how it's used now.